llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1777 @@
+	{
+		return fromBB;
+	}
+	enum VisitingPolicy { NORMAL, REMOVE_VALUES};
+	template <VisitingPolicy policy>
+	llvm::BasicBlock* visitBasicBlock(llvm::BasicBlock* BB, llvm::BasicBlock* from=nullptr);
+	explicit PartialInterpreter(std::unique_ptr<llvm::Module> M)
+		: llvm::Interpreter(std::move(M), /*preExecute*/false)
+	{
+	}
+	BitMask getBitMask(const llvm::Value* V) const
+	{
+		if (isa<ConstantData>(V))
+			return BitMask::ALL;
+
+		if (isa<Constant>(V))
+			return BitMask::ALIGNED8;
+
+		if (stronglyKnownBits.count(V))
+			return stronglyKnownBits.at(V);
+
+		return BitMask::NONE;
+	}
+	bool isValueComputed(const llvm::Value* V) const
+	{
+		if (isValueComputedConstant(V))
+			return true;
+		if (isa<Argument>(V))
+		{
+			return computed.count(V);
+		}
+		if (isa<Instruction>(V))
+		{
+			return computed.count(V);
+		}
+		if (isa<BasicBlock>(V))
+			return true;
+		return false;
+	}
+	bool areOperandsComputed(const llvm::Instruction& I)
+	{
+		if (isa<SelectInst>(I))
+		{
+			if (isValueComputed(I.getOperand(0)))
+			{
+				GenericValue V = getOperandValue(I.getOperand(0), getLastStack());
+				if (V.IntVal == 0u)
+				{
+					if (isValueComputed(I.getOperand(2)))
+						return true;
+				}
+				else
+				{
+					if (isValueComputed(I.getOperand(1)))
+						return true;
+				}
+
+			}	
+		}
+		for (auto& op : I.operands())
+		{
+			if (!isValueComputed(op))
+				return false;
+		}
+		if (const SwitchInst* SI = dyn_cast <SwitchInst>(&I))
+		{
+			if ((getBitMask(SI->getCondition()) != BitMask::ALL) )
+				return false;
+		}
+		if (const BranchInst* SI = dyn_cast <BranchInst>(&I))
+		{
+			if (SI->isConditional())
+			if ((getBitMask(SI->getCondition()) != BitMask::ALL) )
+				return false;
+		}
+		return true;
+	}
+	bool hasToBeSkipped(llvm::Instruction& I) //TODO:const
+	{
+		if (isa<UnreachableInst>(I))
+			return true;
+		if (isa<LandingPadInst>(I)) //LandingPadInst are not handled by Interpreter
+			return true;
+//TODO??	if (isa<StoreInst>(I))
+//			return true;
+
+		if (isa<VAArgInst>(I))
+			return true;
+		if (!areOperandsComputed(I))
+			return true;
+			//ADD THAT RANDOM INSTRUCTIONS ARE NOT EXECUTABLE
+		if (isa<CallBase>(I))
+		{
+//			if (isa<InvokeInst>(I)) //Invokes are not handled by Interpreter
+//				return true;
+		//	return true;
+			CallBase& CB = cast<CallBase>(I);
+			if (!CB.getCalledFunction())
+				return true;
+			{
+		Function* FF = dyn_cast<Function>(CB.getCalledFunction());
+
+	//	llvm::errs() << FF->getName() << "\n";
+		if (FF->isDeclaration())
+			return true;
+
+//willReturn() will not actualy work since we should be fine to infinite loops (we are though?????)
+//TODO!!!
+/*
+		llvm::errs() << "enters\n";
+for (auto& BB : *FF)
+	if (isa<UnreachableInst>(BB.getTerminator()))
+		return true;
+		llvm::errs() << "exites\n";
+
+if (!FF->doesNotThrow())
+	 return true;
+*/
+int i=0;
+				for (auto& op : CB.args())
+			{
+
+
+					if (i >= FF->getFunctionType()->getNumParams())
```

Comment:
Hmm... I suspect vararg might _not_ be a problem. VAInst itself is skipped anyway. For example a varag function whose return value does not depend on the vararg params is still ok